File: others/lanqiao/分教室.cpp
#pragma warning(disable:4996)
#pragma warning(disable:6031)
//#pragma comment(linker, "/STACK:1024000000,1024000000")
//std::ios_base::sync_with_stdio(false);
#define LOCAL
//#define DEBUG
//#define CONSOLE_PAUSE


#ifndef LOCAL
#define NDEBUG
#endif // close assert if not local
#include<iostream>
#include<cstdio>
#include<string>
#include<cstring>
#include<algorithm>
#include<vector>
#include<cctype>
#include<cassert>
using namespace std;
typedef long long ll;
typedef unsigned long long ull;
typedef unsigned int uint;
typedef pair<int, int> Pii;
template <class Int>
inline Int fast_read()
{
	Int x = 0, f = 1; char ch = getchar();
	while (!isdigit(ch)) { if (ch == '-') f = -1; ch = getchar(); }
	while (isdigit(ch)) { x = x * 10 + ch - '0'; ch = getchar(); }
	return x * f;
}
inline int readint() { return fast_read<int>(); }
inline ll readll() { return fast_read<ll>(); }
//------------------------ header end ------------------------

const int maxn = 300 + 5;
const int INF = int(1e5);
int best_ans = INF;
int n;

int colors[maxn];
vector<int> neis[maxn];

bool color_ok(int u, int color)
{
	for (int i = 0; i < int(neis[u].size()); i++)
	{
		int v = neis[u][i];
		if (colors[v] != 0 && colors[v] == color)
			return false;
	}
	return true;
}

void dfs(int u, int cur_num)
{
	if (cur_num >= best_ans)
		return;
	if (u > n)
	{
		best_ans = min(best_ans, cur_num);
		return;
	}
	for (int color = 1; color <= cur_num; color++)
	{
		if (color_ok(u, color))
		{
			colors[u] = color;
			dfs(u + 1, cur_num);
			colors[u] = 0;
		}
	}
	colors[u] = cur_num + 1;
	dfs(u + 1, cur_num + 1);
	colors[u] = 0;
}

void add_bi_edge(int u, int v)
{
	neis[u].push_back(v);
	neis[v].push_back(u);
}

int main()
{
#ifdef LOCAL
	freopen("in.txt", "r", stdin);
	freopen("out.txt", "w", stdout);
#endif

	n = readint();
	int m = readint();
	for (int ct = 1; ct <= m; ct++)
	{
		int u = readint();
		int v = readint();
		add_bi_edge(u, v);
	}
	dfs(1, 0);
	printf("%d", best_ans);

#ifdef CONSOLE_PAUSE
	system("PAUSE");
#endif
	return 0;
}

File: others/lanqiao/分教室.py
import math
import sys
from typing import List, Tuple, Set, Dict
from math import sqrt, ceil, floor, fabs, sin, cos

LOCAL, DEBUG = False, False


#LOCAL = True
# DEBUG = True

def read_list() -> List[str]:
    return input().split(' ')


def read_int_list() -> List[int]:
    return list(map(int, read_list()))


MAXN = 300 + 5
INF = int(1e5)
best_ans = INF
n = 0

colors = [0 for _ in range(MAXN)]
neis: List[List[int]] = [[] for _ in range(MAXN)]


def color_ok(u: int, color: int) -> bool:
    for v in neis[u]:
        if colors[v] == 0:
            continue
        if colors[v] == color:
            return False
    return True


def dfs(u: int, cur_num: int):
    global best_ans
    if cur_num >= best_ans:
        return
    if u > n:
        best_ans = min(best_ans, cur_num)
        return
    for color in range(1, cur_num + 1):
        if color_ok(u, color):
            colors[u] = color
            dfs(u + 1, cur_num)
            colors[u] = 0
    colors[u] = cur_num + 1
    dfs(u + 1, cur_num + 1)
    colors[u] = 0


def add_bi_edge(u: int, v: int):
    neis[u].append(v)
    neis[v].append(u)


def main():
    global n
    n = int(input())
    m = int(input())
    for _ in range(m):
        u, v = read_int_list()
        add_bi_edge(u, v)
    dfs(1, 0)
    print(best_ans, end='')


if __name__ == '__main__':
    sys.setrecursionlimit(int(1e5))
    if LOCAL:
        in_file = open('in.txt', 'r')
        out_file = open('out.txt', 'w')
        sys.stdin = in_file
        sys.stdout = out_file
    main()
